src/artifactContentFileSystemProvider.ts

```diff
@@ -0,0 +1,333 @@
+    /**
+     * The SARIF log the URI was created from.
+     */
+    readonly log: string;
+
+    /**
+     * The original file name specified in the location object of the SARIF.
+     */
+    readonly artifactUri: string;
+
+    /**
+     * The run index in the SARIF log.
+     */
+    readonly runIndex: number;
+
+    /**
+     * The artifact index in the SARIF log (relative to the run).
+     */
+    readonly artifactIndex: number;
+}
+
+/**
+ * Use to return the artifact contents and the data parsed from the embedded artifact URI.
+ * @see EmbeddedContentFileSystemProvider.getArtifactInformationFromUri
+ */
+interface ArtifactInformationFromUri {
+    /**
+     * The artifact.
+     */
+    artifact: sarif.Artifact;
+
+    /**
+     * The artifact contents.
+     */
+    contents: sarif.ArtifactContent;
+
+    /**
+     * The information parsed from the embedded URI.
+     */
+    parsedUriData: ParsedUriData;
+
+    /**
+     * The SARIF log file.
+     */
+    log: sarif.Log;
+}
+
+/**
+ * A file system provider that handles embedded content.
+ */
+export class ArtifactContentFileSystemProvider implements vscode.FileSystemProvider, vscode.Disposable {
+    private disposables: vscode.Disposable[] = [];
+
+    // The URI scheme for the content.
+    private static ArtifactContentScheme: string = 'sarifArtifactContent';
+
+    private readonly onDidChangeFileEventEmitter: vscode.EventEmitter<vscode.FileChangeEvent[]> = new vscode.EventEmitter<vscode.FileChangeEvent[]>();
+
+    /**
+     * Creates in instance of the embedded content file system provider.
+     */
+    public constructor() {
+        this.disposables.push(this.onDidChangeFileEventEmitter);
+        this.disposables.push(vscode.workspace.registerFileSystemProvider(ArtifactContentFileSystemProvider.ArtifactContentScheme, this, {
+            isCaseSensitive: true,
+            isReadonly: true
+        }));
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public get onDidChangeFile(): vscode.Event<vscode.FileChangeEvent[]> {
+        return this.onDidChangeFileEventEmitter.event;
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public watch(uri: vscode.Uri, options: { recursive: boolean; excludes: string[] }): vscode.Disposable {
+        const parsedUriData: ParsedUriData = ArtifactContentFileSystemProvider.parseUri(uri);
+
+        const watcher: (currentStats: fs.Stats, previousStats: fs.Stats) => void = (currentStats, previousStats) => {
+            this.onDidChangeFileEventEmitter.fire([
+                {
+                    type: vscode.FileChangeType.Changed,
+                    uri: parsedUriData.log
+                }
+            ]);
+        };
+
+        fs.watchFile(parsedUriData.log.fsPath, watcher);
+
+        return {
+            dispose: () => { fs.unwatchFile(parsedUriData.log.fsPath, watcher); }
+        };
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public async stat(uri: vscode.Uri): Promise<vscode.FileStat> {
```

Comment:
Returning dummy values seems to make no difference for our scenarios. Why are implementing this fully?

Reply:
We can remove the size calculation for sure.

VSCode will use the times to determine if it can show a "cached" view (on re-start) of content it may have.

---
In reply to: [411806880](https://github.com/microsoft/sarif-vscode-extension/pull/252#discussion_r411806880) [](ancestors = 411806880)